qiskit/providers/ibmq/job/ibmqjob.py

```diff
@@ -203,17 +205,19 @@ def result(self, timeout: Optional[float] = None, wait: float = 5) -> Result:
             Result object
 
         Raises:
-            JobError: if the job has failed or cancelled, or if there was some
-                unexpected failure in the server.
+            IBMQJobApiError: if the job was cancelled or there was some unexpected failure
+                in the server.
+            IBMQJobFailureError: If the job failed.
         """
         # pylint: disable=arguments-differ
         # pylint: disable=access-member-before-definition,attribute-defined-outside-init
 
         if not self._wait_for_completion(timeout=timeout, wait=wait,
                                          required_status=(JobStatus.DONE,)):
-            message = 'Job was cancelled.' if self._status is JobStatus.CANCELLED \
-                else 'Job has failed. Use job.error_message() to get more details.'
-            raise JobError('Unable to retrieve job result. ' + message)
+            if self._status is JobStatus.CANCELLED:
+                raise IBMQJobApiError('Unable to retrieve job result. Job was cancelled.')
```

Comment:
I think we should have another `IBMQJobInvalidStateError` for cases like this, since it's not really an API error.